app/components/elements/form/form/index.js

```diff
@@ -0,0 +1,106 @@
+import {cloneElement} from 'preact'
+import {path, reduce, equal} from 'wasmuth'
+import {dispatch} from '/store'
+import {
+  formSet as formSetBuilder,
+  formRemove as formRemoveBuilder
+} from '../actions.js'
+import compose from '/util/compose'
+import mapStateToProps from '/util/mapStateToProps'
+import BaseForm from './base'
+
+export default mapStateToProps(
+  (state, {name}) => path(['forms', name], state)
+)(compose({
+  componentWillMount () {
+    this.formSet = formSetBuilder(this.props.name)
+    this.formRemove = formRemoveBuilder(this.props.name)
+  },
+  componentWillUnmount () {
+    !this.props.keepFormDataOnUnmount && dispatch(this.formRemove())
+  },
+  componentWillUpdate ({
+    initialData,
+    validations = {},
+    onSubmit,
+    name,
+    data,
+    initialized,
+    asyncErrors,
+    submit,
+    errors: existingErrors
+  }) {
+    if (!initialized) {
+      dispatch(this.formSet(['initialized'], true))
+      dispatch(this.formSet(['data'], initialData))
+    }
+
+    if (submit) {
+      dispatch(this.formSet(['submit'], false))
+      dispatch(this.formSet(['submitting'], true))
+      onSubmit(data, onSuccess(name), onError(name))
+    }
+
+    const errors = validate(data || {}, validations)
+    if (!equal(errors, existingErrors)) {
+      dispatch(this.formSet(['errors'], errors))
+      asyncErrors && dispatch(this.formRemove(['asyncErrors']))
+    }
+  },
+  render ({formSet, name, children, ...props}) {
+    return BaseForm({
+      onSubmit: ev => ev.preventDefault() || dispatch(formSet(['submit'], true)),
+      children: addFormNameToChildren(children, name),
+      ...props
+    })
+  }
+}))
+
+const validate = (data, validations) =>
```

Comment:
So are validations only passed in to the form? Or can i still have field level validations?

Currently on Qee, I have an Input element that has validations run based on the type `type=email,password,number` etc.

Reply:
Makes sense to add default validations to certain input types.
Yeah right now all validations are passed in to the Form component